### DISCUSSION TITLE
Update Limits & Gas Schedule for FEVM

### DISCUSSION THREAD

**Discussion by Stebalien** [filecoin-project/FIPs#588] [Enhancements - Technical]
Discussion for https://github.com/filecoin-project/FIPs/pull/573.

With the introduction of user programmable smart contracts, we need to ensure that the gas model accurately reflects the cost of on-chain computation. In the current network, users can only trigger course-grained behaviors by submitting messages to built-in actors. However, with the introduction of FEVM, users will be able to deploy new smart contracts with fine-grained control over FVM execution. If we don't carefully tune the gas model to accurately charge for computation, an attacker could carefully construct a contract to take advantage that and slow down or even stop chain validation.

This FIP proposes a few adjustments to the gas charging schedule. Specifically:

- Adjustments to some syscall gas:
    - State read and write costs.
    - Internal send costs.
    - Hashing, signature verification, and randomness.
- Variable charges for memory initialization, copying, and allocation.
- New charges for state-tree reads and writes.
- New charges for actor address resolution.

This FIP also introduces overall memory limits. However, these memory limits currently only include Wasm memories and table elements (not IPLD blocks, wasm _code_, and other metadata).

Additionally, this FIP introduces a maximum block size limit of 1MiB for all newly created IPLD blocks.

Finally, this FIP reduces the maximum recursive call depth limit from 1025 to 1024 to bring it in-line with the initial intentions of this limit and other blockchain VMs.

**Comment by anorth** [comment]
Thanks for this. I did initially have a comment here, but have removed it as it was based on a mistaken reading.

This looks good so far. Can we expect some benchmark results to indicate the likely cost changes for core network operations like WindowPoSt and doing deals?

**Reply by kaitlin-beegle** [reply]
I believe this has been addressed by the [data modeling](https://filecoinproject.slack.com/archives/C01EU76LPCJ/p1676036239499679) recently completed by CEL; I've also requested that this info be added to the FIPs repo as supplementary documentation. 

It will be up to CEL and Core Devs to flag anamolies, etc., post-implementation.
